openvisualizer/motehandler/moteprobe/moteprobe.py

```diff
@@ -1,175 +1,34 @@
-                    tester.set_num_test_pkt(1)
-                    tester.set_timeout(2)
-                    tester.test(blocking=True)
-                    if tester.get_stats()['numOk'] >= 1:
-                        mote_ports.append((port, baud))
-                        break
-            except serial.SerialException as e:
-                log.warning('{} {}'.format(e, port))
-            except Exception as e:
-                log.error(e)
-            finally:
-                probe.close()
-                probe.join()
-
-    # log
-    log.success("discovered following serial-port(s): {0}".format(['{0}@{1}'.format(s[0], s[1]) for s in mote_ports]))
-
-    return mote_ports
-
-
-# ============================ class ===========================================
-
-class OpentestbedMoteFinder(object):
-    OPENTESTBED_RESP_STATUS_TIMEOUT = 10
-
-    def __init__(self, mqtt_broker_address):
-        self.opentestbed_motelist = set()
-        self.mqtt_broker_address = mqtt_broker_address
 
-    def get_opentestbed_motelist(self):
+# ============================ class ===================================
 
-        # create mqtt client
-        mqtt_client = mqtt.Client('FindMotes')
-        mqtt_client.on_connect = self._on_mqtt_connect
-        mqtt_client.on_message = self._on_mqtt_message
-        mqtt_client.connect(self.mqtt_broker_address)
-        mqtt_client.loop_start()
-
-        # wait for a while to gather the response from otboxes
-        log.info("discovering motes in testbed... (waiting for {}s)".format(self.OPENTESTBED_RESP_STATUS_TIMEOUT))
-        time.sleep(self.OPENTESTBED_RESP_STATUS_TIMEOUT)
-
-        # close the client and return the motes list
-        mqtt_client.loop_stop()
-
-        log.info("discovered {0} motes".format(len(self.opentestbed_motelist)))
-
-        return self.opentestbed_motelist
-
-    def _on_mqtt_connect(self, client, userdata, flags, rc):
-
-        log.success("succesfully connected to: {0}".format(self.mqtt_broker_address))
-
-        client.subscribe('opentestbed/deviceType/box/deviceId/+/resp/status')
-
-        payload_status = {'token': 123}
-        # publish the cmd message
-        client.publish(
-            topic='opentestbed/deviceType/box/deviceId/all/cmd/status',
-            payload=json.dumps(payload_status),
-        )
-
-    def _on_mqtt_message(self, client, userdata, message):
-
-        # get the motes list from payload
-        payload_status = json.loads(message.payload)
-
-        for mote in payload_status['returnVal']['motes']:
-            if 'EUI64' in mote:
-                self.opentestbed_motelist.add(mote['EUI64'])
+class MoteProbeNoData(Exception):
+    """ No data received from serial pipe """
+    pass
 
+# ============================ class ===================================
 
 class MoteProbe(threading.Thread):
-    class MoteModes(IntEnum):
-        MODE_SERIAL = 0
-        MODE_EMULATED = 1
-        MODE_IOTLAB = 2
-        MODE_TESTBED = 3
 
```

Comment:
__metaclass__ = abc.ABCMeta
to make this class an abstract base class (see top comment on this file)

The class methods that are implemented in the child classes, (e.g., `_attach`, `_detach`, `_send_data`, ... should be defined here too as follows:
`@abc.abstractmethod`
`_attach(self):`
`   raise NotImplementedError("Should be implemented by child class")`